src/tdastro/astro_utils/snia_utils.py

```diff
@@ -7,6 +8,84 @@
 from tdastro.util_nodes.scipy_random import NumericalInversePolynomialFunc
 
 
+def snia_volumetric_rates(redshift):
+    """
+    SN Ia volumetric rate based on Frohmaier et al. (2019).
+    r_v(z) = r0 * (1+z)^alpha （SNe Ia yr^-1 Mpc^-3 h_70^3）
+    r0 = 2.27+/-0.19e-5
+    alpha = 1.7+/-0.21
+
+    Parameters
+    ----------
+    redshift: `float` or `numpy.ndarray`
+        The redshift of the supernova
+
+    Returns
+    -------
+    rate_vol: `float` or `numpy.ndarray`
+        The volumetric rate of the supernova given the redshift
+    """
+
+    r0 = 2.27e-5
+    alpha = 1.7
+    rate_vol = r0 * np.power(1.0 + redshift, alpha)
+
+    return rate_vol
+
+
+def num_snia_per_redshift_bin(zmin=0.001, zmax=10, znbins=20, solid_angle=None, H0=73.0, Omega_m=0.3):
```

Comment:
I don't really get why do we need bin function. For me it looks like `snia_volumetric_rates(redshift) * dV(redshift)` should be enough to build a redshift distribution to sample from...

Reply:
I haven't thought carefully on how this should be used. I think there are two things that are needed: 1. a redshift distribution (which is likely a function of all host properties) 2. The total number of SNe (which also depends on length, solid angle)